src/librustc_resolve/lib.rs

```diff
@@ -2470,6 +2493,32 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
         self.ribs[TypeNS].pop();
     }
 
+    fn with_tuple_struct_self_ctor_rib<F>(&mut self, self_ty: &Ty, f: F)
+        where F: FnOnce(&mut Resolver)
+    {
+        let variant_def = if self.session.features_untracked().tuple_struct_self_ctor {
+            let base_def = self.def_map.get(&self_ty.id).map(|r| r.base_def());
+            if let Some(Def::Struct(ref def_id)) = base_def {
```

Comment:
This won't work for type aliases in `impl`.
```rust
struct S(u8);
type Alias = S;
impl Alias { ... Self(10) ... }
```

Reply:
Thanks. Test case is added.